part_2/solution.py

```diff
@@ -0,0 +1,861 @@
+print(f"Total Validation Images = {len(val_dataloader)}")
+opt.phase= "train"
+# %% [markdown]
+"""
+## Configure Pix2PixHD GAN and train to predict nuclei from phase.
+Having loaded the data into the model we can now train the Pix2PixHD GAN to predict nuclei from phase. We will use the following hyperparameters to train the model:
+
+"""
+# %%
+# Define the parameters for the Generator.
+opt.ngf = 64  # Number of filters in the generator.
+opt.n_downsample_global = 4  # Number of downsampling layers in the generator.
+opt.n_blocks_global = 9  # Number of residual blocks in the generator.
+opt.n_blocks_local = 3  # Number of residual blocks in the generator.
+opt.n_local_enhancers = 1  # Number of local enhancers in the generator.
+
+# Define the parameters for the Discriminators.
+opt.num_D = 3  # Number of discriminators.
+opt.n_layers_D = 3  # Number of layers in the discriminator.
+opt.ndf = 32  # Number of filters in the discriminator.
+
+# Define general training parameters.
+opt.gpu_ids = [0] # GPU ids to use.
+opt.norm = "instance"  # Normalization layer in the generator.
+opt.use_dropout = ""  # Use dropout in the generator (fixed at 0.2).
+opt.batchSize = 8  # Batch size.
+
+# Create a visualizer to perform image processing and visualization
+visualizer = Visualizer(opt)
+
+
+# Here will first start training a model from scrach however we can continue to train from a previously trained model by setting the following parameters.
+opt.continue_train = False
+if opt.continue_train:
+    iter_path = os.path.join(opt.checkpoints_dir, opt.name, "iter.txt")
+    try:
+        start_epoch, epoch_iter = np.loadtxt(iter_path, delimiter=",", dtype=int)
+    except:
+        start_epoch, epoch_iter = 1, 0
+    print("Resuming from epoch %d at iteration %d" % (start_epoch, epoch_iter))
+else:
+    start_epoch, epoch_iter = 1, 0
+    
+print('------------ Options -------------')
+for k, v in sorted(vars(opt).items()):
+    print('%s: %s' % (str(k), str(v)))
+print('-------------- End ----------------')
+
+# Set the number of epoch to be 1 for demonstration purposes
+opt.n_epochs = 2 # start from 1
+# Initialize the model
+phase2nuclei_model = create_model(opt)
+# Define Optimizers for G and D
+optimizer_G, optimizer_D = (
+    phase2nuclei_model.module.optimizer_G,
+    phase2nuclei_model.module.optimizer_D,
+)
+# %%
+train_model(
+    opt,
+    phase2nuclei_model,
+    visualizer,
+    dataset_train,
+    dataset_val,
+    optimizer_G,
+    optimizer_D,
+    start_epoch,
+    epoch_iter,
+    writer,
+)
+# %% [markdown]
+"""
+<div class="alert alert-info">
+
+## A heads up of what to expect from the training...
+<br>
+</div>
+"""
+# %% [markdown]
+"""
+<div class="alert alert-info">
+
+**Visualise Phase, Fluorescence and Virtual Stain for Validation Examples**<br>
+- We can observe how the performance improves over time using the images tab and the sliding window.
+<br><br>
+</div>
+"""
+# %% [markdown]
+"""
+<div class="alert alert-info">
+
```

Comment:
since you are using mixed between html and markdown these don't render well in VSCode.

Reply:
Take a look at all the markdown cells